src/storage/engine/perf_context.rs

```diff
@@ -0,0 +1,143 @@
+        PerfStatistics {
+            absolute: false,
+            internal_key_skipped_count: other.internal_key_skipped_count
+                - self.internal_key_skipped_count,
+            internal_delete_skipped_count: other.internal_delete_skipped_count
+                - self.internal_delete_skipped_count,
+            block_cache_hit_count: other.block_cache_hit_count - self.block_cache_hit_count,
+            block_read_count: other.block_read_count - self.block_read_count,
+            block_read_byte: other.block_read_byte - self.block_read_byte,
+        }
+    }
+
+    /// Create an instance which stores absolute statistics values, retrieved at creation.
+    fn new(perf_context: &PerfContext) -> PerfStatistics {
+        PerfStatistics {
+            absolute: true,
+            internal_key_skipped_count: perf_context.internal_key_skipped_count() as usize,
+            internal_delete_skipped_count: perf_context.internal_delete_skipped_count() as usize,
+            block_cache_hit_count: perf_context.block_cache_hit_count() as usize,
+            block_read_count: perf_context.block_read_count() as usize,
+            block_read_byte: perf_context.block_read_byte() as usize,
+        }
+    }
+
+    pub fn is_absolute(&self) -> bool {
+        self.absolute
+    }
+}
+
+/// Helper to collect RocksDB's `PerfContext`. This struct records the statistics at creation,
+/// and updates delta to the `collect_target` when it is dropped.
+#[must_use = "You must keep the collector live in the scope, otherwise it won't collect anything!"]
+pub struct PerfContextCollector<'a> {
+    collect_target: &'a mut PerfStatistics,
+    perf_context: PerfContext,
+    /// Absolute statistics values retrieved at creation.
+    perf: PerfStatistics,
+}
+
+impl<'a> PerfContextCollector<'a> {
+    pub fn new(collect_target: &'a mut PerfStatistics) -> PerfContextCollector<'a> {
+        let perf_context = PerfContext::get();
+        let perf = PerfStatistics::new(&perf_context);
+        PerfContextCollector {
+            collect_target,
+            perf_context,
+            perf,
+        }
+    }
+
+    /// Get delta statistics since instance creation.
+    pub fn collect(&self) -> PerfStatistics {
+        let current_perf = PerfStatistics::new(&self.perf_context);
+        self.perf.get_delta(&current_perf)
+    }
+}
+
+impl<'a> Drop for PerfContextCollector<'a> {
+    fn drop(&mut self) {
+        let statistics = self.collect();
+        self.collect_target.add(&statistics);
+    }
+}
+
+/// RocksDB's `PerfContext` is thread local. If we send `PerfContextCollector` to another thread,
+/// we will get incorrect delta values, so it is `!Send` and `!Sync`.
+impl<'a> !Send for PerfContextCollector<'a> {}
+
+impl<'a> !Sync for PerfContextCollector<'a> {}
+
+/// A wrapper of `PerfContextCollector` to deal with optional targets. If the collecting target
+/// is `None`, there will be no overheads.
+pub enum OptionTargetPerfContextCollector<'a> {
```

Comment:
Hm, I'm not sure this is necessary. Maybe you can help me understand why this was added?

Reply:
See FAQ 2. Basically, for one function, we sometimes want it to collect metrics, but sometimes not. So this "OptionTargetPerfContextCollector" enables us to control whether or not to collect without much overhead.